File: backend/src/main/java/com/matriculaonline/service/CursoService.java
package com.matriculaonline.service;

import com.matriculaonline.domain.exception.ResourceNotFoundException;
import com.matriculaonline.domain.model.Curso;
import com.matriculaonline.dto.request.CursoRequest;
import com.matriculaonline.dto.response.CursoResponse;
import com.matriculaonline.dto.response.PageResponse;
import com.matriculaonline.repository.CursoRepository;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
public class CursoService {

    private final CursoRepository cursoRepository;

    public CursoService(CursoRepository cursoRepository) {
        this.cursoRepository = cursoRepository;
    }

    @Transactional
    public CursoResponse criar(CursoRequest request) {
        Curso curso = new Curso();
        curso.setNome(request.nome());
        curso.setDescricao(request.descricao());
        curso.setCargaHoraria(request.cargaHoraria());
        return CursoResponse.fromEntity(cursoRepository.save(curso));
    }

    @Transactional(readOnly = true)
    public PageResponse<CursoResponse> listar(Pageable pageable) {
        return PageResponse.from(cursoRepository.findAll(pageable), CursoResponse::fromEntity);
    }

    @Transactional(readOnly = true)
    public CursoResponse buscarPorUuid(UUID uuid) {
        Curso curso = cursoRepository.findByUuid(uuid)
                .orElseThrow(() -> new ResourceNotFoundException("Curso", uuid.toString()));
        return CursoResponse.fromEntity(curso);
    }

    @Transactional
    public CursoResponse atualizar(UUID uuid, CursoRequest request) {
        Curso curso = cursoRepository.findByUuid(uuid)
                .orElseThrow(() -> new ResourceNotFoundException("Curso", uuid.toString()));
        curso.setNome(request.nome());
        curso.setDescricao(request.descricao());
        curso.setCargaHoraria(request.cargaHoraria());
        return CursoResponse.fromEntity(cursoRepository.save(curso));
    }

    /**
     * Remove o curso. Idempotente: se o recurso já não existir, retorna sem erro (no-op).
     */
    @Transactional
    public void deletar(UUID uuid) {
        cursoRepository.findByUuid(uuid).ifPresent(cursoRepository::delete);
    }
}

File: frontend/src/app/services/curso.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Curso, CursoRequest } from '../models/curso.model';
import { PageResponse } from '../models/page.model';

@Injectable({ providedIn: 'root' })
export class CursoService {
  private readonly apiUrl = '/api/cursos';

  constructor(private http: HttpClient) {}

  listar(page = 0, size = 10, sort = 'nome,asc', q?: string): Observable<PageResponse<Curso>> {
    let params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', sort);
    if (q) {
      params = params.set('q', q);
    }
    return this.http.get<PageResponse<Curso>>(this.apiUrl, { params });
  }

  buscarPorUuid(uuid: string): Observable<Curso> {
    return this.http.get<Curso>(`${this.apiUrl}/${uuid}`);
  }

  criar(request: CursoRequest): Observable<Curso> {
    return this.http.post<Curso>(this.apiUrl, request);
  }

  atualizar(uuid: string, request: CursoRequest): Observable<Curso> {
    return this.http.put<Curso>(`${this.apiUrl}/${uuid}`, request);
  }

  deletar(uuid: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${uuid}`);
  }
}
